### PR TITLE
SectionList Component

### PR DESCRIPTION
**Pull request by lancetipton**
**Tickets**
* [ZEN-498](https://jira.simpleviewtools.com/browse/ZEN-498)
* [ZEN-511](https://jira.simpleviewtools.com/browse/ZEN-511)

> **IMPORTANT**
> This PR should be tested along with [this PR](https://github.com/simpleviewinc/tap-events-force/pull/114)

## Updates - 1/21/2021
* added helper method to get scroll values `getScrollValues`
    * utilized this helper in `useScroll` method
    * we may want to get the scroll values outside of a hook
* `sectionList.native.js`
    * added logic to recalculate active session for this edge case:
        * if the section contents changes without scrolling, we want to make sure we update the activeSection accordingly

* added new `test` case in the `Testing` section of the [evf PR](https://github.com/simpleviewinc/tap-events-force/pull/114)

## Updates - 1/20/2021
- updated useScroll logic to account for when scrolling ends
- updated sectionList ref to be set the entire nativeObject ref rather than just a part of it
- updated `useSectionChangeOnScroll` to do calculations for new section only after scrolling ends

## Updates - 01/13/2021

* Fixed an issue with classNames being arrays vs a string
  * Updated StyleInjector to only join the class names to string when the component is a string
* Cleaned up useStyleCallback
  * Added defaults and extra memoize of styles object

<hr/>

[Semver-Status] minor

## Context

We need to update the Sessions Day navigation to match the updated mock-ups
See this ticket => https://jira.simpleviewtools.com/browse/ZEN-472
 

## Goal

* Utilize react-native sectionList
* On Desktop Web show gray line for section titles
* On Mobile show Day number for section titles
* Should also include the name of the day
* On day toggle, it should scroll to the section for that day
* On Desktop web, the day toggle day number needs to be updated when the gray line that separates days touches the bottom of the day toggle component
 

## Updates

* `repos/keg-components/configs/aliases.json`
  * Added alias for new components 
* `repos/keg-components/src/components/divider/divider.js`
 * Updated to use `forwardRef`
* `repos/keg-components/src/components/list/*`
  * Added `SectionList` component
  * Moved `useScrollClassNames` to it's own hook file so it can be reused between list components
  * Added some example of using the `SectionList` component
* `repos/keg-components/src/hooks/*`
  * Added two reusable hooks ( `useScrollClassNames` and `useScroll` )
  * Allows tracking browser scrolling in web environments
* `repos/keg-components/src/theme/components/list/*`
  * Updated / Added the styles for the `SectionList` component
* `repos/keg-components/src/utils/components/*`
  * `getElementLayout` - Helper for interact with an elements layout and position
  * `scrollList` - Helper for scrolling a list of items or the window

## Testing

* Setup
  * Pull this PR => `keg && keg pr 105`   

**Test 1**
* ~~Run the Keg-Components's docker package => `keg dpg run docker.pkg.github.com/simpleviewinc/keg-packages/keg-components:zen-498-session-list-view`~~
* Use this image instead => `keg consumer pack run docker.pkg.github.com/simpleviewinc/keg-packages/keg-test-consumer:master `
  * Navigate your browser to http://components-zen-498-session-list-view.local.kegdev.xyz/
  * In the Nav on the right, select `Components` => `List` => `SectionList`
  * Ensure the `SectionList` example displays properly

### CODE REVIEW

File: repos/keg-components/src/components/list/sectionList.native.js [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+    return sections.map((section, index) => ({
+      ...section,
+      __kegIndex: get(section, indexBy, index),
+    }))
+  }, [sections])
+}
+
+/**
+ * Helper hook to allow tracking scrolling between section
+ * @param {boolean} doScrolling - Is automatic scrolling turned on
```
Comment (daniel112) [inline repos/keg-components/src/components/list/sectionList.native.js:43]: - `doScrolling` && `listRef `is not part of the param anymore it seems.
- also missing `activeSection` && `setActiveSection` `@param`

File: repos/keg-components/src/components/list/sectionList.native.js [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+/**
+ * Helper hook to allow switching / scrolling to a section when a section change happens
+ * @param {boolean} doScrolling - Is automatic scrolling turned on
+ * @param {function} onSectionChange - Consumer Callback method for when a section changes
+ * @param {number} scrollOffset - Offset the final scroll position by some amount (px)
+ * @param {Object} sectionRefs - React ref of all Section divider components.
+ * @param {Object} listRef - React ref of the SectionList component
+ * @param {Object} isScrollingRef - React ref to track if we are scrolling
+ * @param {number} scrollCooldown - Amount of time to wait before allowing scrolling (ms)
+ *
```
Comment (daniel112) [inline repos/keg-components/src/components/list/sectionList.native.js:112]: missing `setActiveSection` jsdoc

File: repos/keg-components/src/components/list/sectionList.native.js [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+  )
+}
+
+/**
+ * Displays a section header for each section, adding a view to helper with tracking
+ * @param {Object} props
+ * @param {Object} props.index - Position of the section being rendered
+ * @param {Object} props.renderSectionHeader - Callback passed from the consumer
+ * @param {Object} props.sectionRefs - Ref Object to hold refs to the section header Dom Node
+ * @param {Array} props.sections - Groups of items to be displayed in the SectionList component
```
Comment (daniel112) [inline repos/keg-components/src/components/list/sectionList.native.js:181]: looks like this should now be an object `section`?

File: repos/keg-components/src/components/list/sectionList.native.js [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+ * @param {Object} props.styles - Merged custom and theme styles
+ */
+const SectionHeader = props => {
+  const {
+    index,
+    onSectionChange,
```
Comment (daniel112) [inline repos/keg-components/src/components/list/sectionList.native.js:187]: missing jsdoc for `onSectionChange`

File: repos/keg-components/src/components/list/sectionList.native.js [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+    },
+    [ renderItem, onSectionChange ]
+  )
+}
+
+/**
+ * SectionList
+ * @summary Default view component that wraps the React Native View component. All props are optional
+ *
+ * @param {Object} props - see View PropTypes
```
Comment (daniel112) [inline repos/keg-components/src/components/list/sectionList.native.js:275]: `see SectionList PropTypes` typo

File: repos/keg-components/src/components/list/sectionList.native.js [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+      return checkCall(renderItem, { item, onSectionChange })
+    },
+    [ renderItem, onSectionChange ]
+  )
+}
+
+/**
+ * SectionList
+ * @summary Default view component that wraps the React Native View component. All props are optional
+ *
```
Comment (daniel112) [inline repos/keg-components/src/components/list/sectionList.native.js:274]: i think it could be helpful to at least add jsdoc description for the callbacks `onScrollSectionChange` and `onSectionChange` for the props that will be passed to the consumer's callback func.

if we're going with the approach of `see x PropTypes`
apparently we can just add the description there and itll show up in the storybook:
something like this:
- [image](https://user-images.githubusercontent.com/3317835/104960563-b7941f80-5991-11eb-9abf-4e0fbe3e6a6f.png)
- [image](https://user-images.githubusercontent.com/3317835/104960611-cda1e000-5991-11eb-99dd-d4a46fe610b1.png)

**Review by daniel112 — CHANGES_REQUESTED** [review]
component looks good on storybook and all test passed.
there's just some minor cleanup items i noticed

File: repos/keg-components/src/utils/components/scrollList.native.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { exists } from '@keg-hub/jsutils'
+
+/**
+ * Helper method to scroll to an item in a list in a **NATIVE ENVIRONMENT**
+ * @function
+ * @param {object} props
+ * @param {object} [props.top] - Vertical scroll to position (Web Only)
+ * @param {object} [props.left] - Horizontal scroll to position (Web Only)
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/utils/components/scrollList.native.js:8]: no

File: repos/keg-components/src/utils/components/scrollList.native.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ * @param {boolean} [props.animated=true] - Type of scrolling ( auto | smooth )
+ */
+export const scrollList = ({ animated = true, top, left }) => {
+  return listRef?.current?.scrollTo({
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/utils/components/scrollList.native.js:12]: updated

File: repos/keg-components/src/utils/components/scrollList.native.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/utils/components/scrollList.native.js:8]: updated

File: repos/keg-components/src/hooks/useScrollClassName/useScrollClassName.native.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Placeholder hook for Native platforms
+ * <br/>Just returns the passed in ref
+ *
+ * @returns {Object|function|undefeind} - Passed in Ref
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/hooks/useScrollClassName/useScrollClassName.native.js:5]: updated

File: repos/keg-components/src/hooks/useScrollClassName/useScrollClassName.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+const isWeb = getPlatform() === 'web'
+
+/**
+ * Custom hook to update the ScrollView with passed in classNames
+ * <br/>Uses getScrollableNode and getInnerViewNode to get access to the Dom Nodes
+ * @param {Array[string]} defClass - Default class to add to the component
+ * @param {string|Array[string]} className - Class or an array of classes to add to the element
+ * @param {Object|function} ref - Ref object passed to the consuming component
+ *
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/hooks/useScrollClassName/useScrollClassName.js:15]: updated

File: repos/keg-components/src/hooks/useScroll/useScroll.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Custom hook to track the browser windows scroll position
+ * <br/> The onScroll prop should be a function on consistent identity for the throttle to work
+ * <br/> I.E. Wrap with `useCallback` if needed
+ * @function
+ * @example
+ * const [{ scrollX, scrollY }] = useScroll()
+ * @param {function} [onScroll] - Function called when the scroll event fires
+ * @param {number} [amount=30] - Throttle amount for the scrollHandler callback
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/hooks/useScroll/useScroll.js:12]: updated

File: repos/keg-components/src/components/list/sectionList.native.js (hunk 1 above)
Comment (lancetipton) [reply repos/keg-components/src/components/list/sectionList.native.js:43]: updated

File: repos/keg-components/src/components/list/sectionList.native.js (hunk 2 above)
Comment (lancetipton) [reply repos/keg-components/src/components/list/sectionList.native.js:112]: updated

File: repos/keg-components/src/components/list/sectionList.native.js (hunk 3 above)
Comment (lancetipton) [reply repos/keg-components/src/components/list/sectionList.native.js:181]: No, it's an array. What makes you say it's an object?

File: repos/keg-components/src/components/list/sectionList.native.js (hunk 4 above)
Comment (lancetipton) [reply repos/keg-components/src/components/list/sectionList.native.js:187]: updated

File: repos/keg-components/src/components/list/sectionList.native.js (hunk 6 above)
Comment (lancetipton) [reply repos/keg-components/src/components/list/sectionList.native.js:276]: Updated

**Review by mpcarolin — COMMENTED** [review]
Looks good and passed all the testing. I just had one suggestion

File: repos/keg-components/src/components/list/sectionList.native.js (hunk 5 above)
Comment (lancetipton) [reply repos/keg-components/src/components/list/sectionList.native.js]: updated

File: repos/keg-components/src/hooks/useScroll/useScroll.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    throttle(event => {
+      const scrollUpdate = {
+        scrollX: window.pageXOffset,
+        scrollY: window.pageYOffset,
+      }
+
+      onScroll(event, scrollUpdate)
+      setScroll(scrollUpdate)
+    }, amount),
+    [onScroll]
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/hooks/useScroll/useScroll.js:32]: ohh, nice catch. Updating now

File: repos/keg-components/src/components/list/sectionList.native.js (hunk 3 above)
Comment (daniel112) [reply repos/keg-components/src/components/list/sectionList.native.js:182]: I thought this was a typo since there's no `sections` in this function. just `section`
and from the usage on line `192` and `254` it looked like it's a singular item. unless im not understanding it right

File: repos/keg-components/src/hooks/useScroll/useScroll.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-components/src/hooks/useScroll/useScroll.js]: Updated

File: repos/keg-components/src/components/list/sectionList.native.js (hunk 3 above)
Comment (lancetipton) [reply repos/keg-components/src/components/list/sectionList.native.js]: Ya, you are correct. I've updated it

**Review by daniel112 — APPROVED** [review]
looks good

**Review by mpcarolin — APPROVED** [review]
👍🏼
